src/geocoding/geocoder.py

```
@@ -0,0 +1,722 @@
+                self.redis_client.ping()
+                logging.info("Connected to Redis cache")
+            except Exception as e:
+                logging.warning(f"Redis connection failed: {e}")
+                self.redis_client = None
+        else:
+            logging.warning("Redis not available for caching")
+    
+    def _make_key(self, address: str) -> str:
+        """Generate cache key from address."""
+        normalized = address.lower().strip()
+        hash_obj = hashlib.md5(normalized.encode('utf-8'))
```

Ruff (0.12.2)

Probable use of insecure hash functions in `hashlib`: `md5`

(S324)